File: typescript/src/core/periodic-report.ts
/**
 * Shared base for periodic report objects (10-K, 10-Q).
 *
 * Sections (`business`, `riskFactors`, `mda`) are extracted from the
 * document by item number; only sections that were actually found are
 * set, so callers can feature-detect with `in` / `!== undefined`.
 */

import { Financials } from "./financials";

export interface ReportFiling {
  extractItems(): Promise<Record<string, string>>;
  cik: string;
  formType?: string;
  accessionNumber?: string;
  api: { getCompanyFacts(cik: string): Promise<Record<string, any>> };
}

export abstract class PeriodicReport {
  protected readonly filing: ReportFiling;
  public readonly items: Record<string, string>;

  public readonly business?: string;
  public readonly riskFactors?: string;
  public readonly mda?: string;

  protected constructor(filing: ReportFiling, items: Record<string, string>) {
    this.filing = filing;
    this.items = items;

    const sections = (this.constructor as typeof PeriodicReport).SECTION_ITEMS;
    for (const [attribute, itemKey] of Object.entries(sections)) {
      const content = items[itemKey];
      if (content) {
        (this as Record<string, unknown>)[attribute] = content;
      }
    }
  }

  /** mapping of section attribute -> item key in the extracted-items map */
  static SECTION_ITEMS: Record<string, string> = {};

  /** Financial statements backed by XBRL company facts. */
  async financials(): Promise<Financials> {
    return Financials.extract(this.filing);
  }
}

File: python/src/sec_edgar_toolkit/core/periodic_report.py
"""Shared base for periodic report objects (10-K, 10-Q)."""

from __future__ import annotations

import logging
from typing import Dict, Union

from ..parsers.items import TenKItem, TenQItem

logger = logging.getLogger(__name__)


class PeriodicReport:
    """Shared behavior for 10-K / 10-Q report objects."""

    #: mapping of section attribute -> item key in the extracted-items dict
    SECTION_ITEMS: Dict[str, Union[str, TenKItem, TenQItem]] = {}

    def __init__(self, filing) -> None:
        self._filing = filing
        try:
            items = filing.extract_items()
        except Exception as exc:  # pragma: no cover - network/parse failures
            logger.warning("Item extraction failed: %s", exc)
            items = {}
        self._items = items

        # Only expose sections that were actually found, so callers can
        # feature-detect with hasattr().
        for attr, item_key in self.SECTION_ITEMS.items():
            content = items.get(item_key)
            if content:
                setattr(self, attr, content)

    @property
    def items(self) -> Dict[str, str]:
        return self._items

    @property
    def financials(self):
        from .financials import Financials

        return Financials.extract(self._filing)
